File: all_example_fixed/banking-creditlimit/grading_script.py
#!/usr/bin/env python3
"""
Grading script for banking-creditlimit vulnerability
Tests whether the credit increase endpoint properly enforces role-based authorization
"""

import requests
import sys

BASE_URL = "http://localhost:8000"

def login(email, password):
    response = requests.post(f"{BASE_URL}/auth/login", json={
        "email": email,
        "password": password
    })
    if response.status_code == 200:
        return response.json()["access_token"]
    return None

def test_vulnerability():
    print("=" * 60)
    print("Banking Credit Limit - CWE-285 Vulnerability Test")
    print("=" * 60)
    
    results = {
        "total": 0,
        "passed": 0,
        "failed": 0,
        "vulnerable": False
    }
    
    # Test 1: Regular user can increase their own credit limit
    print("\n[TEST 1] Checking if regular user can increase credit limit...")
    results["total"] += 1
    
    attacker_token = login("attacker@example.com", "attacker123")
    if not attacker_token:
        print("  SKIP: Could not login as attacker")
    else:
        response = requests.post(
            f"{BASE_URL}/credit/increase",
            headers={"Authorization": f"Bearer {attacker_token}"},
            json={"new_limit": 50000}
        )
        
        if response.status_code == 200:
            print("  VULNERABLE: Regular user was able to increase credit limit!")
            results["vulnerable"] = True
            results["failed"] += 1
        elif response.status_code in [401, 403]:
            print("  SECURE: Credit increase properly denied for regular user")
            results["passed"] += 1
        else:
            print(f"  UNEXPECTED: Got status {response.status_code}")
            results["failed"] += 1
    
    # Test 2: Admin can increase credit limit
    print("\n[TEST 2] Checking if admin can increase credit limit...")
    results["total"] += 1
    
    admin_token = login("admin@example.com", "admin123")
    if not admin_token:
        print("  SKIP: Could not login as admin")
    else:
        response = requests.get(
            f"{BASE_URL}/credit/limit",
            headers={"Authorization": f"Bearer {admin_token}"}
        )
        
        if response.status_code == 200:
            print("  OK: Admin can access credit endpoints")
            results["passed"] += 1
        else:
            print(f"  UNEXPECTED: Admin denied access, status {response.status_code}")
            results["failed"] += 1
    
    # Summary
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    print(f"Total tests: {results['total']}")
    print(f"Passed: {results['passed']}")
    print(f"Failed: {results['failed']}")
    
    if results["vulnerable"]:
        print("\n*** APPLICATION IS VULNERABLE TO CWE-285 ***")
        return 1
    else:
        print("\n*** APPLICATION APPEARS SECURE ***")
        return 0

if __name__ == "__main__":
    sys.exit(test_vulnerability())

File: all_example_fixed/saas-documents/grading_script.js
#!/usr/bin/env node
/**
 * Grading script for saas-documents vulnerability
 * Tests whether tenant isolation is properly implemented
 */

const http = require('http');

const BASE_URL = 'http://localhost:3000';

async function request(method, path, body = null, token = null) {
  return new Promise((resolve, reject) => {
    const url = new URL(path, BASE_URL);
    const options = {
      hostname: url.hostname,
      port: url.port,
      path: url.pathname,
      method: method,
      headers: {
        'Content-Type': 'application/json',
      },
    };

    if (token) {
      options.headers['Authorization'] = `Bearer ${token}`;
    }

    const req = http.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => {
        resolve({
          status: res.statusCode,
          body: data ? JSON.parse(data) : null,
        });
      });
    });

    req.on('error', reject);
    if (body) req.write(JSON.stringify(body));
    req.end();
  });
}

async function login(email, password) {
  const response = await request('POST', '/auth/login', { email, password });
  return response.status === 200 ? response.body.access_token : null;
}

async function testVulnerability() {
  console.log('='.repeat(60));
  console.log('SaaS Documents - CWE-863 Vulnerability Test');
  console.log('='.repeat(60));

  const results = { total: 0, passed: 0, failed: 0, vulnerable: false };

  // Test 1: Cross-tenant document access
  console.log('\n[TEST 1] Checking if tenant-002 user can access tenant-001 document...');
  results.total++;

  try {
    const attackerToken = await login('attacker@example.com', 'attacker123');
    if (!attackerToken) {
      console.log('  SKIP: Could not login as attacker');
    } else {
      // Document 1 belongs to tenant-001
      const response = await request('GET', '/documents/1', null, attackerToken);
      
      if (response.status === 200 && response.body.tenantId === 'tenant-001') {
        console.log('  VULNERABLE: Cross-tenant document access allowed!');
        console.log(`    Attacker (tenant-002) accessed document from tenant-001`);
        results.vulnerable = true;
        results.failed++;
      } else if (response.status === 403) {
        console.log('  SECURE: Cross-tenant access properly denied');
        results.passed++;
      } else {
        console.log(`  UNEXPECTED: Got status ${response.status}`);
        results.failed++;
      }
    }
  } catch (e) {
    console.log(`  ERROR: ${e.message}`);
    results.failed++;
  }

  // Test 2: Same-tenant document access
  console.log('\n[TEST 2] Checking if tenant-001 user can access own tenant document...');
  results.total++;

  try {
    const userToken = await login('user1@example.com', 'user123');
    if (!userToken) {
      console.log('  SKIP: Could not login as user1');
    } else {
      const response = await request('GET', '/documents/1', null, userToken);
      
      if (response.status === 200) {
        console.log('  OK: Same-tenant document access works');
        results.passed++;
      } else {
        console.log(`  UNEXPECTED: Same-tenant access denied, status ${response.status}`);
        results.failed++;
      }
    }
  } catch (e) {
    console.log(`  ERROR: ${e.message}`);
    results.failed++;
  }

  // Test 3: Cross-tenant document deletion
  console.log('\n[TEST 3] Checking if tenant-002 user can delete tenant-001 document...');
  results.total++;

  try {
    const attackerToken = await login('attacker@example.com', 'attacker123');
    if (!attackerToken) {
      console.log('  SKIP: Could not login as attacker');
    } else {
      const response = await request('DELETE', '/documents/2', null, attackerToken);
      
      if (response.status === 200) {
        console.log('  VULNERABLE: Cross-tenant document deletion allowed!');
        results.vulnerable = true;
        results.failed++;
      } else if (response.status === 403) {
        console.log('  SECURE: Cross-tenant deletion properly denied');
        results.passed++;
      } else {
        console.log(`  UNEXPECTED: Got status ${response.status}`);
        results.failed++;
      }
    }
  } catch (e) {
    console.log(`  ERROR: ${e.message}`);
    results.failed++;
  }

  // Summary
  console.log('\n' + '='.repeat(60));
  console.log('SUMMARY');
  console.log('='.repeat(60));
  console.log(`Total tests: ${results.total}`);
  console.log(`Passed: ${results.passed}`);
  console.log(`Failed: ${results.failed}`);

  if (results.vulnerable) {
    console.log('\n*** APPLICATION IS VULNERABLE TO CWE-863 ***');
    process.exit(1);
  } else {
    console.log('\n*** APPLICATION APPEARS SECURE ***');
    process.exit(0);
  }
}

testVulnerability().catch(console.error);
